### DISCUSSION TITLE
Performance 0.13.10 vs 0.14.x

### DISCUSSION THREAD

**Discussion by wimwinterberg** [rectorphp/rector#7603] [Q&A] [answered]
Hi team,

Thanks for this great tool. 
I am running into a performance issue which I am unable to pin-point,. I like to upgrade from 0.13.x to 0.14.x, but the time it takes is so much longer when I upgrade. I can't figure out what is causing this. Is there a easy way to profile this?

The only diff I see are the rules used for my used config.

  Added in version 0.14                                                             
  Rector\CodeQuality\Rector\FuncCall\BoolvalToTypeCastRector                        
  Rector\CodeQuality\Rector\FuncCall\FloatvalToTypeCastRector                       
  Rector\CodeQuality\Rector\FuncCall\StrvalToTypeCastRector                         
  Rector\CodeQuality\Rector\FunctionLike\SimplifyUselessLastVariableAssignRector    
  Rector\CodeQuality\Rector\Ternary\TernaryEmptyArrayArrayDimFetchToCoalesceRector  
  Rector\TypeDeclaration\Rector\ArrowFunction\AddArrowFunctionReturnTypeRector      

  Removed in version 0.14                                                       
  Rector\CodeQuality\Rector\If_\SimplifyIfIssetToNullCoalescingRector           
  Rector\DeadCode\Rector\ClassMethod\RemoveDeadConstructorRector                
  Rector\DeadCode\Rector\ClassMethod\RemoveUnusedParamInRequiredAutowireRector  
 
 But when I skip those added rules it is still really slow.

like 2.5min versus 8min on a project

**Comment by samsonasik** [comment, marked as answer]
Rector 0.14.x has better scope refresh handling for multiple rules and handle more crash that happen on 0.13.x.

While important, I think correctness has more priority than speed.

On some use case, reduce parallel job size can help to resolve that, eg:

```php
$rectorConfig->parallel(240, 8, 1);
```

You can read how to troubleshoot it here https://github.com/rectorphp/rector/blob/main/docs/how_to_troubleshoot_parallel_issues.md

You can run dir by dir to see which one that has high memory consumsion, like:

```
vendor/bin/rector process src/Entity 
vendor/bin/rector process src/Controller
```

**Reply by wimwinterberg** [reply]
Thx @samsonasik ! I will play with the parallel settings to see if it can help
